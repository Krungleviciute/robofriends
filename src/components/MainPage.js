import React, { Component } from "react";
import CardList from "../components/CardList";
import SearchBox from "../components/SearchBox";
import "./MainPage.css";
import Scroll from "../components/Scroll";

import Header from "../components/Header";

// Using class component so it would be possible to test instances of the component
export class MainPage extends Component {
  componentDidMount() {
    this.props.onRequestRobots();
  }

  filterRobots = (robotsData, searchText) => {
    return robotsData.filter((robot) => {
      return robot.name.toLowerCase().includes(searchText.toLowerCase());
    });
  };

  render() {
    const { onSearchChange, isPending, robots, searchField } = this.props;

    return (
      <div className="tc">
        <Header />
        <SearchBox searchChange={onSearchChange} />
        <Scroll>
          {isPending ? (
            <h2 id="loading">Loading</h2>
          ) : (
            <CardList robots={this.filterRobots(robots, searchField)} />
          )}
        </Scroll>
      </div>
    );
  }
}

export default MainPage;

// ------------------- Hooks Example --------------------
// const MainPage = ({
//   searchField,
//   onSearchChange,
//   robots,
//   isPending,
//   onRequestRobots,
// }) => {
//   useEffect(() => {
//     onRequestRobots();
//     // eslint-disable-next-line react-hooks/exhaustive-deps
//   }, []);

//   const filterRobots = (robotsData, text) =>
//     robotsData.filter((robot) => {
//       return robot.name.toLowerCase().includes(text.toLowerCase());
//     });

//   return isPending ? (
//     <h1 className="tc">Loading ...</h1>
//   ) : (
//     <div className="tc">
//       <Header />
//       <SearchBox searchChange={onSearchChange} />
//       <Scroll>
//         <CardList robots={filterRobots(robots, searchField)} />
//       </Scroll>
//     </div>
//   );
// };
// export default MainPage;
