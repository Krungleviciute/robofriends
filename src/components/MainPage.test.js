import React from "react";
import { create } from "react-test-renderer";
import CardList from "./CardList";
import MainPage from "./MainPage";

const mockRobots = [
  {
    id: 1,
    name: "Leanne Graham",
    username: "Bret",
    email: "Sincere@april.biz",
  },
  {
    id: 2,
    name: "Ervin Howell",
    username: "Antonette",
    email: "Shanna@melissa.tv",
  },
];

let wrapper;
let wrapper2;
beforeEach(() => {
  const mockProps = {
    onRequestRobots: jest.fn(),
    robots: [],
    searchField: "",
    isPending: false,
  };

  wrapper = create(<MainPage {...mockProps} />);
});

it("Should render main page without crashing", () => {
  expect(wrapper.toJSON()).toMatchSnapshot();
});

it("should filter robots corectly", () => {
  expect(wrapper.getInstance().filterRobots([], "")).toEqual([]);
  expect(wrapper.getInstance().filterRobots(mockRobots, "z")).toEqual([]);
  expect(wrapper.getInstance().filterRobots(mockRobots, "e")).toEqual(
    mockRobots
  );
  expect(wrapper.getInstance().filterRobots(mockRobots, "a")).toEqual([
    mockRobots[0],
  ]);
});

it("should test when isPending is false/true", () => {
  const mockProps2 = {
    onRequestRobots: jest.fn(),
    robots: [],
    searchField: "",
    isPending: true,
  };

  wrapper2 = create(<MainPage {...mockProps2} />);
  const testInstance = wrapper2.root;
  expect(testInstance.findByType("h2").props.id).toBe("loading");
  // Checks with original wrapper when pending is false and should show cardlist component
  expect(wrapper.root.findByType(CardList)).toBeTruthy();
});
