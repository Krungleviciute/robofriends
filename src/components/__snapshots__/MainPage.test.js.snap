// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render main page without crashing 1`] = `
<div
  className="tc"
>
  <div>
    <h1
      className="f2"
    >
      RoboFriends
    </h1>
  </div>
  <div>
    <input
      className="pa3 ba br3 bg-washed-green"
      placeholder="Search Robots"
      type="search"
    />
  </div>
  <div
    className="scroller"
    style={
      Object {
        "height": "800px",
        "overflowY": "scroll",
      }
    }
  >
    <div />
  </div>
</div>
`;
